homeassistant/components/mqtt/water_heater.py

```diff
@@ -0,0 +1,308 @@
+            value_templates = {
+                key: config.get(CONF_VALUE_TEMPLATE) for key in VALUE_TEMPLATE_KEYS
+            }
+        for key in VALUE_TEMPLATE_KEYS & config.keys():
+            value_templates[key] = config[key]
+        self._value_templates = {
+            key: MqttValueTemplate(
+                template,
+                entity=self,
+            ).async_render_with_possible_json_value
+            for key, template in value_templates.items()
+        }
+
+        self._command_templates = {}
+        for key in COMMAND_TEMPLATE_KEYS:
+            self._command_templates[key] = MqttCommandTemplate(
+                config.get(key), entity=self
+            ).async_render
+
+        support = WaterHeaterEntityFeature(0)
+        if (self._topic[CONF_TEMP_STATE_TOPIC] is not None) or (
+            self._topic[CONF_TEMP_COMMAND_TOPIC] is not None
+        ):
+            support |= WaterHeaterEntityFeature.TARGET_TEMPERATURE
+
+        if (self._topic[CONF_MODE_STATE_TOPIC] is not None) or (
+            self._topic[CONF_MODE_COMMAND_TOPIC] is not None
+        ):
+            support |= WaterHeaterEntityFeature.OPERATION_MODE
+
+        self._attr_supported_features = support
+
+    def _prepare_subscribe_topics(self) -> None:
+        """(Re)Subscribe to topics."""
+        topics: dict[str, dict[str, Any]] = {}
+
+        @callback
+        def handle_mode_received(
+            msg: ReceiveMessage, template_name: str, attr: str, mode_list: str
+        ) -> None:
+            """Handle receiving listed mode via MQTT."""
+            payload = self.render_template(msg, template_name)
+
+            if payload not in self._config[mode_list]:
+                _LOGGER.error("Invalid %s mode: %s", mode_list, payload)
+            else:
+                setattr(self, attr, payload)
+                get_mqtt_data(self.hass).state_write_requests.write_state_request(self)
+
+        @callback
+        @log_messages(self.hass, self.entity_id)
+        def handle_current_mode_received(msg: ReceiveMessage) -> None:
+            """Handle receiving mode via MQTT."""
+            handle_mode_received(
+                msg,
+                CONF_MODE_STATE_TEMPLATE,
+                "_attr_current_operation",
+                CONF_MODE_LIST,
+            )
+
+        self.add_subscription(
+            topics, CONF_MODE_STATE_TOPIC, handle_current_mode_received
+        )
+
+        self.prepare_subscribe_topics(topics)
+
+    async def async_set_operation_mode(self, operation_mode: str) -> None:
+        """Set new operation mode."""
+        payload = self._command_templates[CONF_MODE_COMMAND_TEMPLATE](operation_mode)
+        await self._publish(CONF_MODE_COMMAND_TOPIC, payload)
+
+        if self._optimistic or self._topic[CONF_MODE_STATE_TOPIC] is None:
+            self._attr_current_operation = operation_mode
+            self.async_write_ha_state()
+
+    async def async_set_temperature(self, **kwargs: Any) -> None:
```

Comment:
Not sure about the method names but I think this is cleaner than leaving unimplemented abstract methods and dealing with operation mode typing

Reply:
If the method names were not correct, you would have trouble getting the service calls working. Seems fine to me here.